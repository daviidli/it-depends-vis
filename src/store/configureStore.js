/* eslint-disable no-underscore-dangle */
import { createBrowserHistory } from 'history';
import { applyMiddleware, compose, createStore } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createRootReducer from '../reducers';

export const history = createBrowserHistory();

const configureStore = preloadedState => createStore(
	createRootReducer(history),
	preloadedState,
	compose(
		applyMiddleware(routerMiddleware(history)),
		window.__REDUX_DEVTOOLS_EXTENSION__ ? window.__REDUX_DEVTOOLS_EXTENSION__() : f => f
	)
);

export default configureStore;
