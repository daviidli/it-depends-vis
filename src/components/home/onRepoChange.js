import axios from 'axios';
import { pipe, prop, identity } from 'ramda';
import {
	fork, attemptP, mapRej, map
} from 'fluture/index';
import { push } from 'connected-react-router';
import store from '../../store/store';
import { setData, setError } from '../../actions/actions';
import { dispatch, dispatchThunk } from '../../utils/reduxUtils';
import routes from '../../constants/routes.json';

const goHome = dispatchThunk(store, push(routes.HOME));
const goToVis = dispatchThunk(store, push(routes.VIS));

const dispatchData = pipe(prop('data'), setData, dispatch(store));
const dispatchError = pipe(setError, dispatch(store));

const putRequest = repo => attemptP(() => axios.put(`http://localhost:8080/init?url=${repo}`));

const onRepoChange = pipe(
	putRequest,
	map(pipe(dispatchData, goToVis)),
	mapRej(pipe(dispatchError, goHome)),
	fork(identity)(identity)
);

export default onRepoChange;
