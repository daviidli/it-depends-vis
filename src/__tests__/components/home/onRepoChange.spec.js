import axios from 'axios';
import store from '../../../store/store';
import onRepoChange from '../../../components/home/onRepoChange';

jest.mock('../../../store/store');
jest.mock('axios');

describe('onRepoChange', () => {
	beforeEach(() => {
		jest.resetAllMocks();
	});

	it('should call setData with correct payload', done => {
		store.dispatch.mockImplementation(jest.fn());
		store.getState.mockImplementation(() => ({ repo: 'repo1' }));
		axios.put.mockImplementation(jest.fn().mockImplementation(() => Promise.resolve({ data: 'foo', notData: 'bar' })));

		onRepoChange();
		setImmediate(() => {
			expect(store.dispatch).toHaveBeenCalledTimes(2);
			expect(store.dispatch.mock.calls).toMatchSnapshot();
			done();
		});
	});

	it('should fail and call goHome', done => {
		store.dispatch.mockImplementation(jest.fn());
		store.getState.mockImplementation(() => ({ repo: 'repo1' }));
		axios.put.mockImplementation(jest.fn().mockImplementation(() => Promise.reject(new Error('error'))));

		onRepoChange();
		setImmediate(() => {
			expect(store.dispatch).toHaveBeenCalledTimes(3);
			expect(store.dispatch.mock.calls).toMatchSnapshot();
			done();
		});
	});
});
