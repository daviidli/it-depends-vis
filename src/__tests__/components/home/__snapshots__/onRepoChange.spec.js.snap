// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onRepoChange should call setData with correct payload 1`] = `
Array [
  Array [
    Object {
      "data": "foo",
      "type": "SET_DATA",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "args": Array [
          "/vis",
        ],
        "method": "push",
      },
      "type": "@@router/CALL_HISTORY_METHOD",
    },
  ],
]
`;

exports[`onRepoChange should fail and call goHome 1`] = `
Array [
  Array [
    Object {
      "error": [Error: error],
      "type": "SET_ERROR",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "args": Array [
          "/",
        ],
        "method": "push",
      },
      "type": "@@router/CALL_HISTORY_METHOD",
    },
  ],
]
`;
